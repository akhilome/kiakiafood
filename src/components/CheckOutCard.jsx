import React from 'react';
import PropTypes from 'prop-types';

const CheckOutCard = ({
  id = 0, foodName = '', foodPrice = 0, buttonCallback,
}) => (
  <div data-foodid={id} className="food-card-checkout">
    <div className="food-card-checkout__details">
      <h4>{foodName}</h4>
      <p>
â‚¦
        {foodPrice}
      </p>
    </div>
    <button type="button" onClick={buttonCallback}>
      Remove
    </button>
  </div>
);

CheckOutCard.propTypes = {
  id: PropTypes.number.isRequired,
  foodName: PropTypes.string.isRequired,
  foodPrice: PropTypes.number.isRequired,
  buttonCallback: PropTypes.func.isRequired,
};

export default CheckOutCard;
