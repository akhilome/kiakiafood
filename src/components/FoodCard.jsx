import React from 'react';
import PropTypes from 'prop-types';

const FoodCard = ({
  foodImage, foodName, id, foodPrice, isAdmin,
}) => (
  <div className="food-card">
    <img className="food-image" src={foodImage} alt={foodName} />
    <div className="food-details">
      <div data-foodid={id} className="food-details__legend">
        <h4>{foodName}</h4>
        <p>
â‚¦
          {foodPrice.toLocaleString()}
        </p>
      </div>
      <div className="food-details__action">
        <button type="button">{isAdmin ? 'delete' : 'buy'}</button>
      </div>
    </div>
  </div>
);

FoodCard.propTypes = {
  foodImage: PropTypes.string.isRequired,
  foodName: PropTypes.string.isRequired,
  foodPrice: PropTypes.number.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  id: PropTypes.number.isRequired,
};

export default FoodCard;
