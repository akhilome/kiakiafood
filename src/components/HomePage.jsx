import React from 'react';
import { Link } from 'react-router-dom';

const HomePage = () => (
  <section className="hero">
    <h2>Quell Your Hunger in Minutes</h2>
    <Link to="/menu">
      <button type="button" className="btn-primary">
        Get Food ‚ü∂
      </button>
    </Link>
  </section>
);

export default HomePage;
