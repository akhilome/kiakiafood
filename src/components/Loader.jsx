import React from 'react';

const Loader = () => (
  <div className="loader">
    <img src="https://i.imgur.com/S39IaLt.gif" alt="loading gif" />
  </div>
);

export default Loader;
