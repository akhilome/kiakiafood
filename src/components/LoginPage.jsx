import React from 'react';
import { Link } from 'react-router-dom';

const LoginPage = () => {
  return (
    <div>
      This is the login page! ğŸ”¥ See <Link to="/signup">signup</Link>
    </div>
  );
};

export default LoginPage;
