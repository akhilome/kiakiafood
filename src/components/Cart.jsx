import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import CheckOutCard from './CheckOutCard';
import { removeFromCart, checkout } from '../actions';

class Cart extends Component {
  onCheckoutClick = async () => {
    const { cart, checkout: placeOrder, history } = this.props;
    const foodItems = Object.keys(cart).map(Number);
    await placeOrder(foodItems);
    const {
      fetching: { fetching, errorMessage },
    } = this.props;
    if (!fetching && !errorMessage) return history.push('/order-history');
    return undefined;
  };

  onRemoveClick = (foodId) => {
    const { removeFromCart: removeItemFromCart } = this.props;
    removeItemFromCart(foodId);
  };

  render() {
    const { cart } = this.props;

    const cartItems = Object.keys(cart)
      .map(Number)
      .map(item => ({
        id: item,
        foodName: cart[item].foodName,
        foodPrice: cart[item].foodPrice,
      }));

    return (
      <div className="wrapper">
        <h2>Items To Be Purchased</h2>
        {cartItems.length ? (
          <Fragment>
            <section className="container checkout">
              {cartItems.map(item => (
                <CheckOutCard
                  key={item.id}
                  id={item.id}
                  foodName={item.foodName}
                  foodPrice={item.foodPrice}
                  buttonCallback={() => this.onRemoveClick(item.id)}
                />
              ))}
            </section>
            <div className="order-price">
              <p>
                <strong>total price</strong>
: â‚¦
                <span id="order-price">
                  {cartItems
                    .map(item => item.foodPrice)
                    .reduce((a, b) => a + b)
                    .toLocaleString()}
                </span>
              </p>
            </div>
            <button
              type="button"
              className="btn-primary"
              id="checkout"
              onClick={this.onCheckoutClick}
            >
              Checkout
            </button>
          </Fragment>
        ) : (
          <section className="container checkout">
            <div>No Items in cart</div>
          </section>
        )}
      </div>
    );
  }
}

Cart.propTypes = {
  cart: PropTypes.instanceOf(Object).isRequired,
  removeFromCart: PropTypes.func.isRequired,
  checkout: PropTypes.func.isRequired,
  history: PropTypes.oneOfType([PropTypes.object, PropTypes.func]).isRequired,
  fetching: PropTypes.instanceOf(Object).isRequired,
};

const mapStateToProps = state => ({
  cart: state.cart,
  fetching: state.fetching,
});

export default connect(
  mapStateToProps,
  { removeFromCart, checkout },
)(Cart);
