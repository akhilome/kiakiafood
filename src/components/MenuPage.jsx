import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import FoodCard from './FoodCard';
import { getMenu, addToCart } from '../actions';

export class MenuPage extends Component {
  componentDidMount() {
    const { getMenu: getFoodMenu } = this.props;
    getFoodMenu();
  }

  onButtonClick(foodId, foodName, foodPrice) {
    const { isLoggedIn, history, addToCart: addFoodToCart } = this.props;
    if (!isLoggedIn) return history.push('/login');
    const foodDetails = { foodId, foodName, foodPrice };
    addFoodToCart(foodDetails);
    return undefined;
  }

  render() {
    const { menu: foods } = this.props;
    return (
      <div className="wrapper">
        <h2>Available Food Items</h2>
        <section className="container food-menu">
          {foods.map(food => (
            <FoodCard
              key={food.id}
              foodImage={food.food_image}
              foodName={food.food_name}
              foodPrice={food.price}
              id={food.id}
              isAdmin={false}
              buttonCallback={() => this.onButtonClick(food.id, food.food_name, food.price)}
            />
          ))}
        </section>
      </div>
    );
  }
}

MenuPage.defaultProps = {
  isLoggedIn: false,
};

MenuPage.propTypes = {
  getMenu: PropTypes.func.isRequired,
  addToCart: PropTypes.func.isRequired,
  menu: PropTypes.instanceOf(Array).isRequired,
  history: PropTypes.oneOfType([PropTypes.object, PropTypes.func]).isRequired,
  isLoggedIn: PropTypes.bool,
};

const mapStateToProps = state => ({
  menu: state.menu,
  isLoggedIn: state.user.isLoggedIn,
});

export default connect(
  mapStateToProps,
  { getMenu, addToCart },
)(MenuPage);
