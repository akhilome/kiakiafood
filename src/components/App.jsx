import '../index.css';
import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import PropTypes from 'prop-types';
import { checkAuthStatus, getCart } from '../actions';
import HomePage from './HomePage';
import LoginPage from './LoginPage';
import Signup from './SignupPage';
import Menu from './MenuPage';
import CartPage from './Cart';
import NotFoundPage from './NotFoundPage';
import Loader from './Loader';
import Nav from './Nav';

export class App extends Component {
  async componentDidMount() {
    const { checkAuthStatus: checkUserAuthStatus, getCart: getCartItems } = this.props;
    await checkUserAuthStatus();
    const { isLoggedIn } = this.props;
    if (isLoggedIn) getCartItems();
  }

  render() {
    const { loading } = this.props;
    return (
      <Router>
        <Fragment>
          {loading && <Loader />}
          <Nav />
          <Switch>
            <Route path="/" exact component={HomePage} />
            <Route path="/login" component={LoginPage} />
            <Route path="/signup" component={Signup} />
            <Route path="/menu" component={Menu} />
            <Route path="/cart" component={CartPage} />
            <Route component={NotFoundPage} />
          </Switch>
        </Fragment>
      </Router>
    );
  }
}

App.propTypes = {
  checkAuthStatus: PropTypes.func.isRequired,
  getCart: PropTypes.func.isRequired,
  isLoggedIn: PropTypes.bool,
  loading: PropTypes.bool,
};

App.defaultProps = {
  isLoggedIn: false,
  loading: false,
};

const mapStateToProps = state => ({
  isLoggedIn: state.user.isLoggedIn,
  loading: state.fetching.fetching,
});

export default connect(
  mapStateToProps,
  { checkAuthStatus, getCart },
)(App);
