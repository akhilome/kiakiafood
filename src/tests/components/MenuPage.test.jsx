import React from 'react';
import { mount } from 'enzyme';

import { MenuPage } from '../../components/MenuPage';

describe('<MenuPage />', () => {
  const props = {
    getMenu: jest.fn(),
    menu: [
      {
        id: 1,
        food_image: 'couple.png',
        food_name: '40 Lake',
        price: 102,
      },
    ],
    addToCart: jest.fn(),
    history: { push: jest.fn() },
  };

  const buyButtonSpy = jest.spyOn(MenuPage.prototype, 'onButtonClick');

  it('should add item to cart if user logged in', () => {
    const wrapper = mount(<MenuPage {...{ ...props, isLoggedIn: true }} />);

    const buyButton = wrapper.find('.food-details__action > button');
    buyButton.simulate('click');
    expect(buyButtonSpy).toHaveBeenCalled();
    expect(props.addToCart).toHaveBeenCalledWith({
      foodId: 1,
      foodName: '40 Lake',
      foodPrice: 102,
    });
  });

  it('should not add item to cart if user is not logged in', () => {
    const wrapper = mount(<MenuPage {...{ ...props, isLoggedIn: false }} />);

    const buyButton = wrapper.find('.food-details__action > button');
    buyButton.simulate('click');
    expect(buyButtonSpy).toHaveBeenCalled();
    expect(props.history.push).toHaveBeenCalled();
  });

  it('should render correctly', () => {
    const wrapper = mount(<MenuPage {...props} />);

    expect(props.getMenu).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });
});
