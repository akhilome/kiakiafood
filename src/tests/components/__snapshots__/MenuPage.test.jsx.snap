// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuPage /> should render correctly 1`] = `
<MenuPage
  addToCart={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "foodId": 1,
            "foodName": "40 Lake",
            "foodPrice": 102,
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  getMenu={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  history={
    Object {
      "push": [MockFunction] {
        "calls": Array [
          Array [
            "/login",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    }
  }
  isLoggedIn={false}
  menu={
    Array [
      Object {
        "food_image": "couple.png",
        "food_name": "40 Lake",
        "id": 1,
        "price": 102,
      },
    ]
  }
>
  <div
    className="wrapper"
  >
    <h2>
      Available Food Items
    </h2>
    <section
      className="container food-menu"
    >
      <FoodCard
        buttonCallback={[Function]}
        foodImage="couple.png"
        foodName="40 Lake"
        foodPrice={102}
        id={1}
        isAdmin={false}
        key="1"
      >
        <div
          className="food-card"
        >
          <img
            alt="40 Lake"
            className="food-image"
            src="couple.png"
          />
          <div
            className="food-details"
          >
            <div
              className="food-details__legend"
              data-foodid={1}
            >
              <h4>
                40 Lake
              </h4>
              <p>
                â‚¦
                102
              </p>
            </div>
            <div
              className="food-details__action"
            >
              <button
                onClick={[Function]}
                type="button"
              >
                buy
              </button>
            </div>
          </div>
        </div>
      </FoodCard>
    </section>
  </div>
</MenuPage>
`;
